### PR TITLE
Make CodedOutputStream a trait, with a default implementation for Writer

### LINKED ISSUE #186: unnecessary buffer in CodedOutputStream
Hi @stepancheg 

I note that CodedOutputStream uses a buffer to cache write first, this may be good for some Write like TcpStream or File, but I think it is not necessary for Vector or other Write which already supports memory buffer. 

Maybe we can remove the buffer and  only use Write here, if user wants to encode the protobuf to a TcpStream or File, he can use BufWrite outside manually.

### PR DESCRIPTION
**Pull request by danburkert**
This commit makes `CodedOutputStream` a trait, which makes it easier to use, and more efficient. `WithCodedOutputStream` is no longer needed, and some of the write options on Message have been collapsed. This is a breaking change.

I think something similar could be done for `CodedInputStream` with a bit more work, if you are interested I can try sending a PR for that as well.

fixes #186

### CODE REVIEW

**Comment by danburkert** [comment]
Forgot the motivation - I've been running a personal branch of `rust-protobuf` for close to a year with this and a few other changes.  The reason is that is that I'm using async networking (mio and now tokio), which means messages are already explicitly buffered on the write path.  The internal buffers in `CodedOutputStream` hurt performance measurably in this scenario due to the overhead of double buffering and associated extra allocations.

**Comment by siddontang** [comment]
Hi @danburkert 

How do you handle CodedInputStream? The results in our benchmark and production use  both show that decode is more slow and costs more CPU than encode.

**Comment by danburkert** [comment]
@siddontang I haven't done it yet, so I'm not 100% sure it's possible.  The tricky part, from what I can tell, is figuring out how to thread the limit through the decoding methods, since it can't be stored in a field as it is currently.

**Comment by danburkert** [comment]
I believe the failing test is a backwards compatibility check, which this commit will obviously not pass.

**Comment by stepancheg** [comment]
I've create two constructors: [from Vec](5773985563e7ceb5efca1ffb3ea3421f79e83356) and [from bytes](035d82c80bdbfb55ecdfae97c7a321620e267842). Explanation is [there](https://github.com/stepancheg/rust-protobuf/pull/194#issuecomment-274768135).

Is this issue still actual?

**Comment by danburkert** [comment]
@stepancheg I don't think so; I'll leave more explanation in the linked comment.

**Comment by danburkert** [comment]
@stepancheg I've updated the PR to remove the generic type params.  Additionally I've made a few more changes:

* varint encoding now takes advantage of the optimization you suggested in #194.
* fixed length int/float encoding now use a similar optimization by reusing the implementation in the `byteorder` crate. This should also fix encoding on big endian platforms, which is currently broken.

I've looked more at the `CodedInputStream` side of things, and a similar change should be quite straightforward.

File: src/lib/stream.rs [hunk 1]
```diff
@@ -665,276 +622,109 @@ impl<'a> WithCodedInputStream for &'a [u8] {
...
-        let mut temp = value;
-        loop {
-            if (temp & !0x7Fu64) == 0 {
-                self.write_raw_byte(temp as u8)?;
-                break;
-            } else {
-                self.write_raw_byte(((temp & 0x7F) | 0x80) as u8)?;
-                temp >>= 7;
-            }
+    fn write_raw_varint64(&mut self, mut value: u64) -> ProtobufResult<()> {
```
Comment (stepancheg) [inline src/lib/stream.rs]: I've pushed it.

File: src/lib/stream.rs [hunk 2]
```diff
@@ -665,276 +622,109 @@ impl<'a> WithCodedInputStream for &'a [u8] {
...
     }
 
-    pub fn write_raw_little_endian32(&mut self, value: u32) -> ProtobufResult<()> {
-        self.write_raw_byte(((value) & 0xFF) as u8)?;
-        self.write_raw_byte(((value >> 8) & 0xFF) as u8)?;
-        self.write_raw_byte(((value >> 16) & 0xFF) as u8)?;
-        self.write_raw_byte(((value >> 24) & 0xFF) as u8)?;
-        Ok(())
+    fn write_raw_little_endian32(&mut self, value: u32) -> ProtobufResult<()> {
+        self.write_u32::<LittleEndian>(value).map_err(From::from)
```
Comment (stepancheg) [inline src/lib/stream.rs]: Done: f6f1eec85cd510b4238108e4aad0f22f8f5aa07b

**Comment by stepancheg** [comment]
@danburkert 

I'm not yet ready to make it dynamic dispatch to `Write` by default (and break backward compat. and so on).

However, I'm thinking about customizing generated code similarly to how it is done [in gogo/protobuf](https://github.com/gogo/protobuf/blob/master/extensions.md).

For example, if you want to generate dynamically-dispatched `Write` calls, you would have to specify an option like `write_no_buffer`.

When this option is specified, generated code will implement functions using `Write` instead of `CodedOutputStream`.

Option could be specified either in per-file, e. g.

```
option (rustproto.write_no_buffer) = true;

message FooBar { ... }
```

or by using alternative generator, e. g. `protoc-gen-rustnobuf` instead of `protoc-gen-rust` (invoking `protoc --rustnobuf_out` instead of `protoc --rust_out`).

Do you think it can work?

**Comment by danburkert** [comment]
OK fair enough.  It's not convenient for me to change the source .protos (they live in another project), so that particular workaround wouldn't be useful for me.

**Comment by stepancheg** [comment]
@danburkert 

That was a recommendation from one of protobuf author from google (cannot find a link now): configure your build system to insert instructions at the beginning of the file before running protoc.

I also proposed to create an alternative executable `protoc-gen-rustnobuf`.

Before that I proposed `BufWrite` trait.

None of my proposals work for you, you simply insist on single solution which breaks compatibility, and you are not open for discussion and compromise.

So I think there's nothing I can do. I'm sorry. I'm closing the issue for now.

**Comment by danburkert** [comment]
Ok, let me know when you start working on 2.0.

**Pull request closed without merging**